components/script/dom/indexeddb/idbtransaction.rs

```diff
@@ -176,27 +249,178 @@ impl IDBTransaction {
 
     pub fn add_request(&self, request: &IDBRequest) {
         self.requests.borrow_mut().push(Dom::from_ref(request));
-        // Increase the number of outstanding requests so that we can detect when
-        // the transaction is allowed to finish.
-        self.pending_request_count
-            .set(self.pending_request_count.get() + 1);
+        self.inc_pending_requests();
     }
 
     /// Must be called by an `IDBRequest` when it finishes (either success or
     /// error). When the last pending request has completed and the transaction
     /// is no longer active, the `"complete"` event is dispatched and any
     /// associated `IDBOpenDBRequest` `"success"` event is fired afterwards.
     pub fn request_finished(&self) {
+        self.dec_pending_requests();
+        if self.aborted.get() {
+            self.maybe_finish_after_abort();
+        } else {
+            self.maybe_finish_successfully();
+        }
+    }
+
+    pub(crate) fn set_upgrade_metadata(&self, old_version: u64) {
+        self.upgrade_old_version.set(Some(old_version));
+    }
+
+    pub(crate) fn initiate_abort(&self, error: Error, can_gc: CanGc) {
+        // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+        // A transaction can be aborted at any time before it is finished,
+        // even if the transaction isn’t currently active or hasn’t yet started.”
+        // Abort can be initiated any time before the transaction finishes.
+        // https://w3c.github.io/IndexedDB/#transaction-lifecycle
+        if self.finished.get() || self.aborted.get() {
+            return;
+        }
+
+        if self.error.get().is_none() {
+            if let Ok(exception) = create_dom_exception(&self.global(), error, can_gc) {
+                self.error.set(Some(&exception));
+            }
+        }
+
+        self.aborted.set(true);
+        let _ = self
+            .get_idb_thread()
+            .send(IndexedDBThreadMsg::AbortTransaction {
+                origin: self.global().origin().immutable().clone(),
+                name: self.db.get_name().to_string(),
+                transaction: self.serial_number,
+            });
+        // https://w3c.github.io/IndexedDB/#abort-a-transaction
+        // When a transaction is aborted the implementation must undo (roll back)
+        // any changes that were made to the database during that transaction.”
+        self.queue_abort_pending_requests();
```

Comment:
So the spec is weird but I think you should only send the message above, do the abort on the backend, and then continue in script for this and the below.